Add Keycloak OAuth login

nickclyde opened the pull request:
# PULL REQUEST

## Summary

This adds an OAuth login flow to the app. Instead of using the sign in page, the sign in button in the Header now redirects to Keycloak. When you login with Keycloak successfully, it redirects back to the app at `/query`.

To test this locally, be sure to `npm install` to get auth.js, then you can run Keycloak via the configuration I added to the docker compose dev file. EG `docker compose -f docker-compose-dev.yaml up keycloak`. ~~Once it's running, you'll need to go to http://localhost:8080 and create a client named `query-connector`. Then you'll need to grab the client secret and set the following variables in `.env`:~~

I updated the PR to automatically import the needed client configuration now you can just grab the needed environment variables from `.env.sample`, and login with `qc-admin` and `QcDev2024!`.

More info on this page: https://authjs.dev/getting-started/providers/keycloak

## Related Issue

Fixes https://linear.app/skylight-cdc/issue/QUE-22/spike-for-supporting-key-cloak

robertandremitchell reviewed (commented):
<img width="1132" alt="Screenshot 2024-11-14 at 11 22 30 AM" src="https://github.com/user-attachments/assets/9e38ce56-cd34-4a07-ba70-bd6169af8399">


I made it to this page but did not see a client secret? Is there an additional step we need to do here in terms of the client authentication?

m-goggins commented:
Is there supposed to be an option to create a user on the localhost:8080 page? I can only see a login:

![image](https://github.com/user-attachments/assets/dc26af7a-2aab-4822-b7ee-b8e8a9b58746)

nickclyde commented:
@robertandremitchell @m-goggins I updated the PR to automatically import the needed client configuration now you can just grab the needed environment variables from `.env.sample`, and login with `qc-admin` and `QcDev2024!`. I also updated the README!

m-goggins commented on query-connector/README.md:
````diff
@@ -12,7 +12,26 @@ The Query Connector app can be run using Docker (or any other OCI container runt
+
+```
+Username: qc-admin
+Password: QcDev2024!
````
Maybe add a line after this about spinning up the app via `npm run dev or npm run dev-win`  and checking that you are able to log in at localhost:3000? In my case, I saw the sign in button appear and then disappear and to check that I keycloak worked, I checked that I was able to get to the my queries page and click log out.

nickclyde replied:
Added!

m-goggins replied:
Looks good!

m-goggins commented on query-connector/.env.sample:
```diff
@@ -0,0 +1,6 @@
+AUTH_KEYCLOAK_ID=query-connector
+AUTH_KEYCLOAK_SECRET=ZG3f7R1J3qIwBaw8QtttJnJMinpERQKs
+AUTH_KEYCLOAK_ISSUER=http://localhost:8080/realms/master
+ERSD_API_KEY=
```
```suggestion
ERSD_API_KEY=
UMLS_API_KEY=
```

m-goggins commented on query-connector/README.md:
```diff
@@ -12,7 +12,26 @@ The Query Connector app can be run using Docker (or any other OCI container runt
+
+Adjust your `DATABASE_URL` as needed.
+
+Add your API key as an environment variable called `ERSD_API_KEY` in the `.env` file so that it can be accessed when running the Query Connector app.
```
```suggestion
Add your API keys as an environment variables called `ERSD_API_KEY` and `UMLS_API_KEY` in the `.env` file so that they can be accessed when running the Query Connector app.
```

m-goggins commented on query-connector/README.md:
````diff
@@ -12,7 +12,26 @@ The Query Connector app can be run using Docker (or any other OCI container runt
+docker compose up -f docker-compose-dev.yaml up Keycloak
+```
+
+To login via Keycloak, use the following credentials:
````
```suggestion
To login via Keycloak, make sure your `.env` is updated using `cp` command above and use the following credentials to login at `localhost:8080` after spinning up the container:
```

m-goggins commented on query-connector/README.md:
````diff
@@ -12,7 +12,26 @@ The Query Connector app can be run using Docker (or any other OCI container runt
+#### Running Keycloak for Authentication
+
+```
+docker compose up -f docker-compose-dev.yaml up Keycloak
````
```suggestion
docker compose up -f docker-compose-dev.yaml up keycloak
```

m-goggins reviewed (approved):
Very cool that you were able to automate more steps! I added a few comments to clarify and I think we need one more line about what we should be looking for to verify that auth worked.
